### PR TITLE
feat(desktop): embed VS Code as a native tab (beta)

### PR DESCRIPTION
**Pull request by omarpr**
## Description

Concrete implementation for the conversation started in #3530. Embeds Microsoft's `code serve-web` inside a native Superset tab so you can browse a worktree, read diffs, and do light editing without spinning up a separate IDE window.

Opening this as **Draft / WIP** on purpose — this is my first Superset contribution and I'd much rather shape the approach with feedback than push for a merge. Happy to split, refactor, or rethink any layer of this.

## Related Issues

Implements #3530.

## Type of Change

- [x] New feature
- [x] Refactor (internal tab/pane plumbing to accept a new `vscode` pane type)

## Approach

- **One shared server, many panes.** `VscodeManager` in the main process owns a single `code serve-web` child. Each pane gets its own `WebContentsView` loading `http://127.0.0.1:<port>/?folder=<worktreePath>`. All panes share the same browser origin, so VS Code's IDB-backed UI state (open editors, sidebar width, command palette history, walkthrough completion) persists across panes and across app restarts.
- **Pinned port with orphan reclamation.** The server binds to a fixed port on every launch so the browser origin stays stable. A PID file at `<appData>/superset-vscode/server.pid` lets the next main-process lifetime reclaim a `code serve-web` child that survived a dev hot-reload or Electron crash. Falls back cleanly to an ephemeral port if the port really is taken.
- **Workspace-independent disk paths.** Server data dir and browser session storage both anchor at `app.getPath("appData")/superset-vscode/*` — intentionally NOT `userData`, because dev builds override the app name per workspace and `userData` derives from the app name. The stable path keeps User settings, extensions, and IDE state shared across every worktree.
- **Focus hand-off.** The embedded view forwards OS-level focus/blur events so the renderer's global hotkey handler stops intercepting shortcuts while the IDE owns keyboard input.
- **Overlay reconciliation.** `WebContentsView` composites above all HTML, so Radix / cmdk / sonner overlays would be hidden behind it. The renderer watches for intersecting overlays, snapshots the view via `capturePage`, paints the PNG as the container's CSS background, then hides the native view — so the IDE image doesn't flash to the window background color when overlays open.

## Features

- **Beta flag.** On by default (opt-out). Toggle under Settings → Behavior → "VS Code (beta)". When off, the "+ VS Code" button is hidden, no VS Code pane can be created, and the `VscodePane` module (already `React.lazy`'d) is never loaded.
- **One VS Code tab per workspace.** Two panes loading the same `?folder=<worktreePath>` would fight over the same IDB bucket. The "+" tab menu's VS Code button is disabled with a tooltip ("VS Code is already open in this workspace") when a worktree already has a VS Code tab open. Different worktrees work in parallel — one VS Code tab per worktree, not one globally.
- **Freeze-frame overlay backdrop.** Before hiding the native view for an overlay, snapshot the current frame and paint it as the container background. Clears on show. Eliminates the flash-to-background-color most of the time (see known issues for the residual).

## Known issues / open questions

- **~20–30 ms flicker on overlay hide.** The capture + hide round-trip isn't instantaneous, so on fast-opening Radix / cmdk overlays a brief corner flash is still visible. Significantly reduced from the naive "hide on intersect" approach but not eliminated. Open to ideas here — could imagine an alternative where we paint the backdrop up-front on a looser debounce, or use a transparent overlay window instead of hiding the view.
- **First-launch bootstrap for existing dev sessions.** If a prior instance left an orphaned `code serve-web` behind AND no PID file exists yet (pre-upgrade), the very first launch after merging this will still fall through to an ephemeral port once. A single manual `pkill -f "code serve-web"` solves it. Every subsequent launch reclaims automatically. Mostly dev-only; production users get the reclaim path on every launch after install.
- **CLI variant.** Tested with the stock `code` CLI from the VS Code desktop installer. Forks like Cursor or Code-OSS probably work (same `serve-web` surface) but are untested.
- **Platform coverage.** Developed and tested on macOS. The code paths (`appData` shape, port probe via `net.createServer`, `ps -p <pid> -o command=` for reclaim verification) are written to be portable, but Linux and Windows deserve a fresh pair of eyes before release.
- **Scope.** This is desktop only. `apps/web` is untouched. If there's appetite for an in-browser variant later, the `VscodeManager` abstraction is close to portable.

## Testing

- **Unit tests** live next to their sources in `apps/desktop/src/main/lib/vscode/` — `vscode-server.test.ts`, `vscode-manager.test.ts`, `reclaim-orphan-server.test.ts`, `find-free-port.test.ts`, `check-code-cli.test.ts`. Covers happy paths, orphan reclamation edge cases (stale file, dead process, recycled PID, unparseable contents), port fallback, CLI absence.
- **Run with:** `bun test src/main/lib/vscode/` from `apps/desktop/`.
- **Typecheck:** `bun run typecheck` in `apps/desktop/` passes clean.
- **Manual smoke test:**
  - Beta flag off → VS Code entry points hidden everywhere, no server spawn.
  - Beta flag on → open a VS Code tab, pick a theme, install an extension, quit, relaunch → no walkthrough reprompt, theme still applied, extension still present, sidebar width preserved.
  - Open a second worktree in a different tab → VS Code button enabled there.
  - Try to open a second VS Code tab in the same worktree → button disabled, tooltip appears.
  - Open a Radix dropdown over the IDE → dropdown renders on top, IDE doesn't flash to background color behind it.

## Screenshots

VS Code embedded in a tab

How to open VS Code in a tab

Setting to disable this feature, marked as Beta

## Additional Notes

- **Allow edits from maintainer:** ✅
- Happy to split this into smaller PRs if preferred — e.g. (1) server + router + manager plumbing, (2) pane component + focus/overlay handling, (3) settings flag + one-per-workspace UX. Let me know if that'd make review easier.
- Thanks for a tool that's genuinely become my command center. Very excited to try to give back.

---
## Summary by cubic
Embed VS Code as a native tab (beta) using a single shared `code serve-web` instance so settings, extensions, and UI state persist across tabs and restarts. Requires the `code` CLI on your PATH; the beta is ON by default and can be toggled in Settings → Behavior.

- **New Features**
  - One shared server managed by `VscodeManager`; each pane loads `<sharedUrl>/?folder=<worktreePath>` in a `WebContentsView`.
  - Stable origin and session: pinned port with orphan reclamation and an app‑data–anchored browser session; settings, keybindings, and extensions persist across worktrees and relaunches.
  - VS Code–first UX: host hotkeys pause while the IDE has focus; conflicting menu accelerators removed; Files/Changes sidebar auto‑hides in VS Code tabs.
  - Clean overlays with a freeze‑frame backdrop and rAF‑coalesced view hide/show; tooltip interactions are debounced; missing CLI screen with install link.
  - Beta flag (default ON) in Settings → Behavior → “VS Code (beta)”; UI is lazy‑loaded; disabling prompts to close open VS Code tabs; “Add VS Code” is hidden until the setting finishes loading; one VS Code tab per worktree is enforced; persisted tabs show a “beta is disabled” message instead of starting the server.
  - Security hardening: `WebContentsView` uses `nodeIntegration: false`, `contextIsolation: true`, `sandbox: true`; block navigation to non‑localhost origins and deny all `window.open` calls.

- **Bug Fixes**
  - Removed the Window → Close accelerator and restored the platform‑correct label (macOS: “Close Window”, others: “Close”) so Cmd/Ctrl+W works inside VS Code.
  - Hardened lifecycle: cancel mid‑boot starts, surface navigation errors (await `loadURL`), handle spawn errors, retry if the shared server dies, recreate views, safely stop on pane/window/app close, and update panes on server crash via an `onStatus` subscription.
  - Fixed Behavior settings regressions (duplicate JSX, dead block) and renumbered the DB migration to avoid conflicts.

<sup>Written for commit 34a1818d7d55d80998b4569667021ebc9314089b. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Embedded VS Code panes: open a VS Code instance inside a workspace tab.
  * “Add VS Code” option in the tab menu to create a per-workspace VS Code tab.
  * VS Code (Beta) toggle in Behavior settings with confirmation dialog when disabling.

* **Bug Fixes & Improvements**
  * Automatic visibility management when UI overlays appear, with snapshot fallback.
  * CLI availability detection with user guidance and safe shutdown on app/window close.
  * Per-tab focus and lifecycle handling to avoid duplicate VS Code panes.

### CODE REVIEW

**Comment by omarpr** [comment]
@coderabbitai review

File: apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/TabView/VscodePane/VscodePane.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		{ enabled: !!workspaceId },
+	);
+	const worktreePath =
+		pane?.vscode?.worktreePath ?? workspace?.worktreePath ?? "";
+
+	const { containerRef, phase, errorMessage } = useEmbeddedVscode({
+		paneId,
+		tabId,
+		worktreePath,
+	});
```
[the remark this answers is not in the document]
Comment (omarpr) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/TabView/VscodePane/VscodePane.tsx:54]: But work trees aren't required, right? What if the user wants to open this on their main/master branch?

Check the below type you proof your suggestion. If that's wrong... Did we do a type check?

**Comment by omarpr** [comment]
@coderabbitai review

File: apps/desktop/src/main/lib/vscode/vscode-server.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+				code,
+				signal,
+				reason: signal ? "killed" : "exited",
+				outputTail: this.outputTail,
+			};
+			this.child = null;
+			this.removePidFile();
+			this.emit("exit", event);
+		});
+	}
```
[the remark this answers is not in the document]
Comment (omarpr) [reply apps/desktop/src/main/lib/vscode/vscode-server.ts:125]: Is this still an issue?

File: apps/desktop/src/main/windows/main.ts [hunk 3]
```diff
@@ -130,6 +133,32 @@ export async function MainWindow() {
...
+	// UI state (walkthrough, sidebar width, recently opened, etc.) every launch.
+	const vscodePidFilePath = join(vscodeSharedRoot, "server.pid");
+	const vscodeManager = new VscodeManager({
+		getWindow,
+		serverDataDir: vscodeServerDataDir,
+		browserSessionDir: vscodeBrowserSessionDir,
+		preferredPort: DEFAULT_PREFERRED_VSCODE_PORT,
+		pidFilePath: vscodePidFilePath,
+	});
+	registerVscodeManager(vscodeManager);
```
[the remark this answers is not in the document]
Comment (omarpr) [reply apps/desktop/src/main/windows/main.ts:171]: Is this still an issue?

File: apps/desktop/src/renderer/routes/_authenticated/settings/behavior/components/BehaviorSettings/BehaviorSettings.tsx [hunk 4]
```diff
@@ -159,128 +173,211 @@ export function BehaviorSettings({ visibleItems }: BehaviorSettingsProps) {
...
+	const handleVscodeBetaConfirm = () => {
+		const panes = useTabsStore.getState().panes;
+		for (const [paneId, pane] of Object.entries(panes)) {
+			if (pane.type === "vscode") {
+				removePane(paneId);
+			}
+		}
+		setVscodeBetaEnabled.mutate({ enabled: false });
+		setVscodeBetaConfirmOpen(false);
+	};
```
[the remark this answers is not in the document]
Comment (omarpr) [reply apps/desktop/src/renderer/routes/_authenticated/settings/behavior/components/BehaviorSettings/BehaviorSettings.tsx:191]: Is this still an issue?
